File: SL-Front-End/src/lib/api/auto-trading.ts
import { axiosInstance } from "../axios";

// Request Types
export interface AutoTradingActivateRequest {
  session_id: string;
  initial_capital?: number;
  allocated_capital: number;
}

export interface AutoTradingDeactivateRequest {
  sell_all_positions: boolean;
}

// Response Types
export interface AutoTradingActivateResponse {
  message: string;
  strategy_id: string;
  is_active: boolean;
  activated_at: string;
}

export interface AutoTradingDeactivateResponse {
  message: string;
  strategy_id: string;
  is_active: boolean;
  deactivated_at: string;
  positions_sold: number;
}

export interface LivePositionResponse {
  position_id: string;
  stock_code: string;
  stock_name?: string;
  quantity: number;
  avg_buy_price: number;
  current_price?: number;
  buy_date: string;
  hold_days: number;
  unrealized_profit?: number;
  unrealized_profit_pct?: number;
  selection_reason?: string;
}

export interface LiveTradeResponse {
  trade_id: string;
  trade_date: string;
  trade_time?: string;
  trade_type: string;
  stock_code: string;
  stock_name?: string;
  quantity: number;
  price: number;
  amount: number;
  commission: number;
  tax: number;
  profit?: number;
  profit_rate?: number;
  hold_days?: number;
  selection_reason?: string;
  order_status: string;
}

export interface LiveDailyPerformanceResponse {
  performance_id: string;
  date: string;
  cash_balance: number;
  stock_value: number;
  total_value: number;
  daily_return?: number;
  cumulative_return?: number;
  buy_count: number;
  sell_count: number;
  trade_count: number;
  position_count: number;
}

export interface AutoTradingStrategyResponse {
  strategy_id: string;
  user_id: string;
  simulation_session_id: string;
  is_active: boolean;
  initial_capital: number;
  current_capital: number;
  cash_balance: number;
  per_stock_ratio: number;
  max_positions: number;
  rebalance_frequency: string;
  created_at: string;
  activated_at?: string;
  deactivated_at?: string;
  last_executed_at?: string;
}

export interface AutoTradingStatusResponse {
  strategy: AutoTradingStrategyResponse;
  positions: LivePositionResponse[];
  today_trades: LiveTradeResponse[];
  latest_performance?: LiveDailyPerformanceResponse;
  total_positions: number;
  total_trades: number;
}

export interface AutoTradingExecuteResponse {
  message: string;
  selected_count: number;
  bought_count: number;
  stocks: Array<{
    stock_code: string;
    company_name: string;
    current_price: number;
    market_cap: number;
    per?: number;
    pbr?: number;
  }>;
}

export interface TradeSignalItem {
  stock_code: string;
  stock_name?: string;
  quantity?: number;
  target_weight?: number;
  current_price?: number;
  per?: number;
  pbr?: number;
  metadata?: Record<string, unknown>;
}

export interface RebalancePreviewResponse {
  generated_at: string;
  candidates: TradeSignalItem[];
  note?: string;
}

export interface AutoTradingLog {
  log_id: string;
  event_type: string;
  event_level: string;
  message?: string;
  details?: Record<string, unknown>;
  created_at: string;
}

export interface AutoTradingLogListResponse {
  logs: AutoTradingLog[];
}

export interface AutoTradingRiskAlert {
  type: string;
  severity: string;
  message: string;
  metadata?: Record<string, unknown>;
}

export interface AutoTradingPositionRisk {
  stock_code: string;
  stock_name?: string;
  quantity: number;
  market_value: number;
  avg_buy_price: number;
  current_price: number;
  unrealized_profit: number;
  unrealized_profit_pct: number;
  hold_days: number;
}

export interface AutoTradingRiskSnapshotResponse {
  as_of: string;
  cash_balance: number;
  invested_value: number;
  total_value: number;
  exposure_ratio: number;
  alerts: AutoTradingRiskAlert[];
  positions: AutoTradingPositionRisk[];
}

export interface ExecutionReportRow {
  date: string;
  live_total_value?: number;
  live_daily_return?: number;
  backtest_total_value?: number;
  backtest_daily_return?: number;
  tracking_error?: number;
}

export interface ExecutionReportSummary {
  days: number;
  average_tracking_error?: number;
  cumulative_live_return?: number;
  cumulative_backtest_return?: number;
  realized_vs_expected?: number;
}

export interface AutoTradingExecutionReportResponse {
  strategy_id: string;
  session_id?: string;
  generated_at: string;
  rows: ExecutionReportRow[];
  summary: ExecutionReportSummary;
}

export const autoTradingApi = {
  /**
   * 자동매매 활성화
   */
  activateAutoTrading: async (
    request: AutoTradingActivateRequest,
  ): Promise<AutoTradingActivateResponse> => {
    const response = await axiosInstance.post<AutoTradingActivateResponse>(
      "/auto-trading/activate",
      request,
    );
    return response.data;
  },

  /**
   * 자동매매 비활성화
   */
  deactivateAutoTrading: async (
    strategyId: string,
    request: AutoTradingDeactivateRequest,
  ): Promise<AutoTradingDeactivateResponse> => {
    const response = await axiosInstance.post<AutoTradingDeactivateResponse>(
      `/auto-trading/strategies/${strategyId}/deactivate`,
      request,
    );
    return response.data;
  },

  /**
   * 자동매매 전략 상태 조회
   */
  getAutoTradingStatus: async (
    strategyId: string,
  ): Promise<AutoTradingStatusResponse> => {
    const response = await axiosInstance.get<AutoTradingStatusResponse>(
      `/auto-trading/strategies/${strategyId}/status`,
    );
    return response.data;
  },

  /**
   * 내 자동매매 전략 목록 조회
   */
  getMyAutoTradingStrategies: async (): Promise<
    AutoTradingStrategyResponse[]
  > => {
    const response = await axiosInstance.get<AutoTradingStrategyResponse[]>(
      "/auto-trading/my-strategies",
    );
    return response.data;
  },

  /**
   * 내 자동매매 전략 목록 조회 (서버 사이드)
   */
  getMyAutoTradingStrategiesServer: async (
    token: string,
  ): Promise<AutoTradingStrategyResponse[]> => {
    const axios = (await import("axios")).default;
    // Docker 환경에서는 컨테이너 이름 사용
    const baseURL = process.env.API_BASE_URL?.replace('/api/v1', '') || "http://backend:8000";
    const response = await axios.get<AutoTradingStrategyResponse[]>(
      `${baseURL}/api/v1/auto-trading/my-strategies`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      },
    );
    return response.data;
  },

  /**
   * 자동매매 수동 실행 (테스트용)
   */
  executeAutoTrading: async (
    strategyId: string,
  ): Promise<AutoTradingExecuteResponse> => {
    const response = await axiosInstance.post<AutoTradingExecuteResponse>(
      `/auto-trading/strategies/${strategyId}/execute`,
    );
    return response.data;
  },

  /**
   * 리밸런싱 프리뷰 조회
   */
  getRebalancePreview: async (
    strategyId: string,
  ): Promise<RebalancePreviewResponse> => {
    const response =
      await axiosInstance.get<RebalancePreviewResponse>(
        `/auto-trading/strategies/${strategyId}/rebalance-preview`,
      );
    return response.data;
  },

  /**
   * 전략 이벤트 로그 조회
   */
  getStrategyLogs: async (
    strategyId: string,
    params?: { event_type?: string; limit?: number },
  ): Promise<AutoTradingLogListResponse> => {
    const response = await axiosInstance.get<AutoTradingLogListResponse>(
      `/auto-trading/strategies/${strategyId}/logs`,
      { params },
    );
    return response.data;
  },

  /**
   * 위험 스냅샷 조회
   */
  getRiskSnapshot: async (
    strategyId: string,
  ): Promise<AutoTradingRiskSnapshotResponse> => {
    const response =
      await axiosInstance.get<AutoTradingRiskSnapshotResponse>(
        `/auto-trading/strategies/${strategyId}/risk`,
      );
    return response.data;
  },

  /**
   * 실거래 리포트 조회
   */
  getExecutionReport: async (
    strategyId: string,
    days = 30,
  ): Promise<AutoTradingExecutionReportResponse> => {
    const response =
      await axiosInstance.get<AutoTradingExecutionReportResponse>(
        `/auto-trading/strategies/${strategyId}/execution-report`,
        { params: { days } },
      );
    return response.data;
  },

  /**
   * 포트폴리오 대시보드 조회
   */
  getPortfolioDashboard: async (): Promise<{
    total_assets: number;
    total_return: number;
    total_profit: number;
    active_strategy_count: number;
    total_positions: number;
    total_trades_today: number;
  }> => {
    const response = await axiosInstance.get(`/auto-trading/dashboard`);
    return response.data;
  },

  /**
   * 포트폴리오 대시보드 조회 (서버 사이드)
   */
  getPortfolioDashboardServer: async (token: string): Promise<{
    total_assets: number;
    total_return: number;
    total_profit: number;
    active_strategy_count: number;
    total_positions: number;
    total_trades_today: number;
  }> => {
    const axios = (await import("axios")).default;
    const baseURL = process.env.API_BASE_URL?.replace('/api/v1', '') || "http://backend:8000";
    const response = await axios.get(
      `${baseURL}/api/v1/auto-trading/dashboard`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      },
    );
    return response.data;
  },
};

File: SL-Back-end/app/schemas/auto_trading.py
"""
자동매매 Pydantic 스키마
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, date, time
from decimal import Decimal
from uuid import UUID


# Request Schemas
class AutoTradingActivateRequest(BaseModel):
    """자동매매 활성화 요청"""
    session_id: str = Field(..., description="백테스트 세션 ID")
    initial_capital: Optional[Decimal] = Field(None, description="초기 자본금 (None이면 키움 계좌 잔고 자동 조회)")
    allocated_capital: Decimal = Field(..., description="전략에 할당할 자본금 (원). 여러 전략에 계좌 잔액을 나누어 배분 가능.", gt=0)
    strategy_name: Optional[str] = Field(None, description="자동매매 전략 이름 (미입력시 자동 생성)")


class AutoTradingDeactivateRequest(BaseModel):
    """자동매매 비활성화 요청"""
    sell_all_positions: bool = Field(True, description="보유 종목 전량 매도 여부")
    deactivation_mode: Optional[str] = Field(None, description="비활성화 모드: immediate(즉시), sell_and_deactivate(매도 후 비활성화), scheduled_sell(예약 매도)")


# Response Schemas
class LivePositionResponse(BaseModel):
    """보유 종목 응답"""
    position_id: UUID
    stock_code: str
    stock_name: Optional[str]
    quantity: int
    avg_buy_price: Decimal
    current_price: Optional[Decimal]
    buy_date: date
    hold_days: int
    unrealized_profit: Optional[Decimal]
    unrealized_profit_pct: Optional[Decimal]
    selection_reason: Optional[str]

    class Config:
        from_attributes = True


class LiveTradeResponse(BaseModel):
    """매매 내역 응답"""
    trade_id: UUID
    trade_date: date
    trade_time: Optional[time]
    trade_type: str
    stock_code: str
    stock_name: Optional[str]
    quantity: int
    price: Decimal
    amount: Decimal
    commission: Decimal
    tax: Decimal
    profit: Optional[Decimal]
    profit_rate: Optional[Decimal]
    hold_days: Optional[int]
    selection_reason: Optional[str]
    order_status: str

    class Config:
        from_attributes = True


class LiveDailyPerformanceResponse(BaseModel):
    """일일 성과 응답"""
    performance_id: UUID
    date: date
    cash_balance: Decimal
    stock_value: Decimal
    total_value: Decimal
    daily_return: Optional[Decimal]
    cumulative_return: Optional[Decimal]
    buy_count: int
    sell_count: int
    trade_count: int
    position_count: int

    class Config:
        from_attributes = True


class AutoTradingStrategyResponse(BaseModel):
    """자동매매 전략 응답"""
    strategy_id: UUID
    user_id: UUID
    simulation_session_id: str
    strategy_name: Optional[str] = Field(None, description="자동매매 전략 이름")
    is_active: bool
    initial_capital: Decimal
    allocated_capital: Decimal
    current_capital: Decimal
    cash_balance: Decimal
    per_stock_ratio: Decimal
    max_positions: int
    rebalance_frequency: str
    created_at: datetime
    activated_at: Optional[datetime]
    deactivated_at: Optional[datetime]
    last_executed_at: Optional[datetime]
    scheduled_deactivation: Optional[bool] = Field(None, description="예약 비활성화 여부")
    deactivation_mode: Optional[str] = Field(None, description="비활성화 모드")
    deactivation_requested_at: Optional[datetime] = Field(None, description="비활성화 요청 시각")
    kiwoom_total_eval: Optional[Decimal] = Field(None, description="키움 API 총 평가액")
    kiwoom_total_profit: Optional[Decimal] = Field(None, description="키움 API 평가손익")
    kiwoom_total_profit_rate: Optional[Decimal] = Field(None, description="키움 API 수익률(%)")

    class Config:
        from_attributes = True


class AutoTradingStatusResponse(BaseModel):
    """자동매매 상태 응답"""
    strategy: AutoTradingStrategyResponse
    positions: List[LivePositionResponse]
    today_trades: List[LiveTradeResponse]
    latest_performance: Optional[LiveDailyPerformanceResponse]
    total_positions: int
    total_trades: int


class AutoTradingActivateResponse(BaseModel):
    """자동매매 활성화 응답"""
    message: str
    strategy_id: UUID
    is_active: bool
    activated_at: datetime


class AutoTradingDeactivateResponse(BaseModel):
    """자동매매 비활성화 응답"""
    message: str
    strategy_id: UUID
    is_active: bool
    deactivated_at: datetime
    positions_sold: int


class AutoTradingLogResponse(BaseModel):
    """자동매매 로그 응답"""
    log_id: UUID
    event_type: str
    event_level: str
    message: Optional[str]
    details: Optional[Dict[str, Any]]
    created_at: datetime

    class Config:
        from_attributes = True


class TradeSignalItem(BaseModel):
    """리밸런싱/시그널 항목"""
    stock_code: str
    stock_name: Optional[str]
    quantity: Optional[int]
    target_weight: Optional[float]
    current_price: Optional[float]
    per: Optional[float]
    pbr: Optional[float]
    metadata: Optional[Dict[str, Any]] = None


class RebalancePreviewResponse(BaseModel):
    """리밸런싱 미리보기 응답"""
    generated_at: datetime
    candidates: List[TradeSignalItem]
    note: Optional[str] = None


class AutoTradingLogListResponse(BaseModel):
    """로그 리스트 응답"""
    logs: List[AutoTradingLogResponse]


class AutoTradingRiskAlert(BaseModel):
    """위험 경보"""
    type: str
    severity: str
    message: str
    metadata: Optional[Dict[str, Any]] = None


class AutoTradingPositionRisk(BaseModel):
    """포지션별 위험 정보"""
    stock_code: str
    stock_name: Optional[str]
    quantity: int
    market_value: Decimal
    avg_buy_price: Decimal
    current_price: Decimal
    unrealized_profit: Decimal
    unrealized_profit_pct: Decimal
    hold_days: int


class AutoTradingRiskSnapshotResponse(BaseModel):
    """위험 스냅샷 응답"""
    as_of: datetime
    cash_balance: Decimal
    invested_value: Decimal
    total_value: Decimal
    exposure_ratio: float
    alerts: List[AutoTradingRiskAlert]
    positions: List[AutoTradingPositionRisk]


class ExecutionReportRow(BaseModel):
    """실거래 vs 백테스트 비교 행"""
    date: date
    live_total_value: Optional[Decimal]
    live_daily_return: Optional[Decimal]
    backtest_total_value: Optional[Decimal]
    backtest_daily_return: Optional[Decimal]
    tracking_error: Optional[Decimal]


class ExecutionReportSummary(BaseModel):
    """실행 보고서 요약"""
    days: int
    average_tracking_error: Optional[Decimal]
    cumulative_live_return: Optional[Decimal]
    cumulative_backtest_return: Optional[Decimal]
    realized_vs_expected: Optional[Decimal]


class AutoTradingExecutionReportResponse(BaseModel):
    """실거래 검증 리포트"""
    strategy_id: UUID
    session_id: str
    generated_at: datetime
    rows: List[ExecutionReportRow]
    summary: ExecutionReportSummary


class AutoTradingRiskEnforceResponse(BaseModel):
    """위험 통제 실행 결과"""
    message: str
    actions: List[Dict[str, Any]]


class PortfolioDashboardResponse(BaseModel):
    """포트폴리오 대시보드 응답"""
    total_assets: Decimal  # 총 자산 (자동매매 계좌 전체)
    total_return: Decimal  # 총 수익률 (%)
    total_profit: Decimal  # 총 수익금
    active_strategy_count: int  # 활성 전략 수
    total_positions: int  # 전체 보유 종목 수
    total_trades_today: int  # 오늘 총 매매 건수
    total_allocated_capital: Decimal  # 자동매매에 할당된 총 금액
